src/atomic.rs

```diff
@@ -0,0 +1,878 @@
+    }
+
+    fn txid(&self) -> Uuid {
+        self.txid.expect("txid")
+    }
+
+    fn is_timeout(&self) -> bool {
+        Utc::now().signed_duration_since(self.started).to_std().expect("duration") > self.timeout
+    }
+
+    fn read_wake_up(&mut self) -> Result<(String, Uuid), Error> {
+        self.bus.as_mut().expect("bus").read_wake_up()
+    }
+
+    fn read_message(&mut self) -> Result<Message, Error> {
+        self.bus.as_mut().expect("bus").read_message()
+    }
+
+    fn write_message(&self, msg: &Message) -> Result<(), Error> {
+        self.bus.as_ref().expect("bus").write_message(msg)
+    }
+
+    fn write_ack(&self) -> Result<(), Error> {
+        self.write_message(&Message::Ack {
+            txid:    self.txid(),
+            serial:  self.serial.clone(),
+            state:   self.state,
+            payload: if let Some(ref signed) = self.signed {
+                Ok(json::to_vec(signed)?)
+            } else if is_terminal(self.state) {
+                Err(Error::AtomicSigned)
+            } else {
+                Ok(Vec::new())
+            }?
+        })
+    }
+}
+
+
+fn parse_signed(payload: &[u8]) -> Result<TufSigned, Error> {
+    Ok(json::from_slice(payload)?)
+}
+
+fn is_valid(from: State, to: State) -> bool {
+    VALID_TRANSITIONS.get(&from).expect("transitions").contains(&to)
+}
+
+fn is_terminal(state: State) -> bool {
+    match state {
+        State::Commit | State::Abort => true,
+        _ => false
+    }
+}
+
+fn is_waiting(err: &Error) -> bool {
+    match *err {
+        Error::Io(ref err) if err.kind() == ErrorKind::TimedOut => true,
+        Error::Io(ref err) if err.kind() == ErrorKind::WouldBlock => true,
+        _ => false
+    }
+}
+
+
+/// Listens for and sends UDP multicast messages.
+pub struct Multicast {
+    wake_up: UdpSocket,
+    message: UdpSocket,
+    wake_addr: SocketAddrV4,
+    msg_addr:  SocketAddrV4,
+}
+
+impl Multicast {
+    pub fn new(wake_addr: SocketAddrV4, msg_addr: SocketAddrV4) -> Result<Self, Error> {
+        Ok(Multicast {
+            wake_up: Multicast::new_socket(wake_addr)?,
+            message: Multicast::new_socket(msg_addr)?,
+            wake_addr: wake_addr,
+            msg_addr:  msg_addr,
+        })
+    }
+
+    fn new_socket(addr: SocketAddrV4) -> Result<UdpSocket, Error> {
+        let any = Ipv4Addr::new(0,0,0,0);
+        let socket = UdpBuilder::new_v4()?
+            .reuse_address(true)?
+            .reuse_port(true)?
+            .bind((any, addr.port()))?;
+        socket.set_broadcast(true)?;
+        socket.set_send_buffer_size(BUFFER_SIZE)?;
+        socket.set_recv_buffer_size(BUFFER_SIZE)?;
+        socket.set_read_timeout(Some(Duration::from_secs(1)))?;
+        socket.join_multicast_v4(&addr.ip(), &any)?;
+        Ok(socket)
+    }
+}
+
+impl Bus for Multicast {
+    fn read_wake_up(&mut self) -> Result<(String, Uuid), Error> {
+        let mut buf = Box::new(vec![0; BUFFER_SIZE]);
```

Comment:
Good point. I tried before to use a fixed buffer but the borrow checker wasn't happy without an owned structure. It should be fine now to put one fixed size buf in the `Multicast` struct though and reference that.